orchestrator/careplanservice/service.go

```diff
@@ -60,51 +67,204 @@ type Service struct {
+	// Creating a resource
+	mux.HandleFunc("POST "+basePath+"/{type}", s.profile.Authenticator(baseUrl, func(httpResponse http.ResponseWriter, request *http.Request) {
+		resourceType := request.PathValue("type")
+		s.handleCreateOrUpdate(request, httpResponse, resourceType, "CarePlanService/Create"+resourceType)
+	}))
+	// Updating a resource
+	mux.HandleFunc("PUT "+basePath+"/{type}/{id}", s.profile.Authenticator(baseUrl, func(httpResponse http.ResponseWriter, request *http.Request) {
+		resourceType := request.PathValue("type")
+		resourceId := request.PathValue("id")
+		s.handleCreateOrUpdate(request, httpResponse, resourceType+"/"+resourceId, "CarePlanService/Update"+resourceType)
+	}))
+	mux.HandleFunc("GET "+basePath+"/{resourcePath...}", s.profile.Authenticator(baseUrl, func(writer http.ResponseWriter, request *http.Request) {
+		// TODO: Authorize request here
+		log.Warn().Msgf("Unmanaged FHIR operation at CarePlanService: %s %s", request.Method, request.URL.String())
+		s.proxy.ServeHTTP(writer, request)
+	}))
 }
 
-func (s Service) RegisterHandlers(mux *http.ServeMux) {
-	proxy := coolfhir.NewProxy(log.Logger, s.fhirURL, basePath, s.transport)
-	baseURL := s.orcaPublicURL.JoinPath(basePath)
-	s.profile.RegisterHTTPHandlers(basePath, baseURL, mux)
-	//
-	// Authorized endpoints
-	//
-	mux.HandleFunc("POST "+basePath+"/Task", s.profile.Authenticator(baseURL, func(writer http.ResponseWriter, request *http.Request) {
-		err := s.handleCreateTask(writer, request)
+func (s *Service) commitTransaction(request *http.Request, tx *coolfhir.TransactionBuilder, resultHandlers []FHIRHandlerResult) (*fhir.Bundle, error) {
+	var txResult fhir.Bundle
+	if err := s.fhirClient.Create(tx.Bundle(), &txResult, fhirclient.AtPath("/")); err != nil {
+		log.Error().Err(err).Msgf("Failed to execute transaction (url=%s)", request.URL.String())
+		// We don't want to log errors from the backing FHIR server for security reasons.
+		return nil, coolfhir.NewErrorWithCode("upstream FHIR server error", http.StatusBadGateway)
+	}
+	var resultBundle fhir.Bundle
+	var notificationResources []any
+	for entryIdx, resultHandler := range resultHandlers {
+		currResult, currNotificationResources, err := resultHandler(&txResult)
 		if err != nil {
-			coolfhir.WriteOperationOutcomeFromError(err, "CarePlanService/CreateTask", writer)
+			return nil, fmt.Errorf("bundle execution succeeded, but couldn't resolve bundle.entry[%d] results: %w", entryIdx, err)
+		}
+		if currResult != nil {
+			resultBundle.Entry = append(resultBundle.Entry, *currResult)
+		}
+		notificationResources = append(notificationResources, currNotificationResources...)
+	}
+
+	for _, notificationResource := range notificationResources {
+		s.notifySubscribers(request.Context(), notificationResource)
+	}
+	return &resultBundle, nil
+}
+
+func (s *Service) proceedTransaction(writer http.ResponseWriter, request *http.Request, resourcePath string, tx *coolfhir.TransactionBuilder) (FHIRHandlerResult, error) {
```

Comment:
Nitpick: It's a bit difficult to understand what this method does by name, maybe something like `ProgressTransaction` or `handleTransaction` would be clearer?

Reply:
changes it to `handleTransactionEntry` and added more comments